vavr/src/main/java/io/vavr/control/Validation.java

```diff
@@ -534,43 +543,45 @@ default void forEach(Consumer<? super T> action) {
     }
 
     /**
-     * Applies a function f to the error of this Validation if this is an Invalid. Otherwise does nothing
+     * Applies a function f to the errors of this Validation if this is an Invalid. Otherwise does nothing
      * if this is a Valid.
      *
-     * @param <U> type of the error resulting from the mapping
-     * @param f   a function that maps the error in this Invalid
+     * @param <U> type of the errors resulting from the mapping
+     * @param f   a function that maps the errors in this Invalid
      * @return an instance of Validation&lt;U,T&gt;
      * @throws NullPointerException if mapping operation f is null
      */
-    default <U> Validation<U, T> mapError(Function<? super E, ? extends U> f) {
+    default <U> Validation<U, T> mapError(Function<Seq<? extends E>, ? extends U> f) {
```

Comment:
contra-variant position:

```java
Function<? super Seq<? super E>, ? extends U>
```

**However**, there is one thing I recognized: We now implicitely have errors of type `Seq<U>` when we have a type `Validation<U, T>`. Isn't it natural then to describe how to map _one_ error? I.e.

```java
default <U> Validation<U, T> mapError(Function<? super E, ? extends U> f) {
    Objects.requireNonNull(f, "f is null");
    if (isInvalid()) {
        return Validation.invalid(f.apply(errors));
    } else {
        return (Validation<U, T>) this;
    }
}
```

I don't know, **please discuss!**

(maybe it would also affect `fold`, `bimap`, ... ?)

Maybe we need also to change the name: `mapErrors` instead of `mapError`.

Reply:
Hi Daniel,
So this was the main decision I made in this change: that all methods that could be used in one fluent block should have the same return type.
The main problem I had when trying to use Vavr validation for nested fields was that sometimes a single error was returned, and sometimes a sequence of errors. This meant that I as a user had to seriously think about how to combine these nested fields: I ended up creating terribly unreadable code, as the business logic (of methods of validation for each field) was inextricably mixed up in low-level logic like fold, bimap, etc.
I created a very simple test, "shouldCreateSuccessNestedCombine()", to illustrate the kind of nested mapping I wanted to simplify. If validation of a set of fields relies on the content of another field, it's very easy to get lost in the code, if you have to map between multiple return 
types. http://www.baeldung.com/vavr mentions this issue.
An example of nested validation is when a customer selects their company type, (which itself is validated), and depending on this type, we either validate their address, or date of birth.
With the change I made, this becomes trivial to do.

Reply:
Hi Amy, your main idea of unifying error to be a sequence is great. I will look into it again today and pull it in.
Because of or work on 1.0.0 currently all is 'in the flow'. Minor subsequent changes may follow...